File: Models/PlannedPlanting.swift
import SwiftData

@Model
final class PlannedPlanting {
    var id: UUID
    var cropName: String
    var variety: String?
    var plannedDate: Date?
    var actualDate: Date?
    var status: PlanStatus
    var notes: String?
    var createdAt: Date
    var updatedAt: Date

    @Relationship(deleteRule: .nullify, inverse: \Garden.plannedPlantings)
    var garden: Garden?

    @Relationship(deleteRule: .nullify, inverse: \GrowingSpace.plannedPlantings)
    var growingSpace: GrowingSpace?

    @Relationship(deleteRule: .nullify, inverse: \Occupancy.plannedPlanting)
    var occupancy: Occupancy?

    @Relationship(deleteRule: .nullify, inverse: \Desire.plannedPlantings)
    var desire: Desire?

    @Relationship(deleteRule: .nullify, inverse: \Seed.plannedPlantings)
    var seed: Seed?

    init(
        cropName: String,
        variety: String? = nil,
        plannedDate: Date? = nil,
        actualDate: Date? = nil,
        status: PlanStatus = .planned,
        notes: String? = nil,
        garden: Garden? = nil,
        growingSpace: GrowingSpace? = nil,
        occupancy: Occupancy? = nil,
        desire: Desire? = nil,
        seed: Seed? = nil
    ) {
        self.id = UUID()
        self.cropName = cropName
        self.variety = variety
        self.plannedDate = plannedDate
        self.actualDate = actualDate
        self.status = status
        self.notes = notes
        self.garden = garden
        self.growingSpace = growingSpace
        self.occupancy = occupancy
        self.desire = desire
        self.seed = seed
        self.createdAt = Date()
        self.updatedAt = Date()
    }

    var displayName: String {
        if let variety = variety, !variety.isEmpty {
            return "\(variety) \(cropName)"
        }
        return cropName
    }
}

enum PlanStatus: String, CaseIterable {
    case planned
    case planted
    case cancelled

    var displayName: String {
        switch self {
        case .planned: return "Planned"
        case .planted: return "Planted"
        case .cancelled: return "Cancelled"
        }
    }
}

File: app/src/main/java/com/soilandsupper/domain/model/PlannedPlanting.kt
package com.soilandsupper.domain.model

import androidx.room.Entity
import androidx.room.ForeignKey
import androidx.room.PrimaryKey

@Entity(
    tableName = "planned_plantings",
    foreignKeys = [
        ForeignKey(
            entity = Garden::class,
            parentColumns = ["id"],
            childColumns = ["gardenId"],
            onDelete = ForeignKey.CASCADE
        ),
        ForeignKey(
            entity = GrowingSpace::class,
            parentColumns = ["id"],
            childColumns = ["growingSpaceId"],
            onDelete = ForeignKey.SET_NULL
        ),
        ForeignKey(
            entity = Occupancy::class,
            parentColumns = ["id"],
            childColumns = ["occupancyId"],
            onDelete = ForeignKey.SET_NULL
        ),
        ForeignKey(
            entity = Desire::class,
            parentColumns = ["id"],
            childColumns = ["desireId"],
            onDelete = ForeignKey.SET_NULL
        ),
        ForeignKey(
            entity = Seed::class,
            parentColumns = ["id"],
            childColumns = ["seedId"],
            onDelete = ForeignKey.SET_NULL
        )
    ]
)
data class PlannedPlanting(
    @PrimaryKey(autoGenerate = true)
    val id: Long = 0,
    val cropName: String,
    val variety: String? = null,
    val plannedDate: Long? = null,
    val actualDate: Long? = null,
    val status: String = PlanStatus.PLANNED.name,
    val notes: String? = null,
    val gardenId: Long? = null,
    val growingSpaceId: Long? = null,
    val occupancyId: Long? = null,
    val desireId: Long? = null,
    val seedId: Long? = null,
    val createdAt: Long = System.currentTimeMillis(),
    val updatedAt: Long = System.currentTimeMillis()
) {
    val displayName: String
        get() = if (!variety.isNullOrBlank()) "$variety $cropName" else cropName
}

enum class PlanStatus {
    PLANNED,
    PLANTED,
    CANCELLED;

    val displayName: String
        get() = when (this) {
            PLANNED -> "Planned"
            PLANTED -> "Planted"
            CANCELLED -> "Cancelled"
        }
}
